contracts/utils/OptionsPurchaseQueue.sol

```diff
@@ -0,0 +1,442 @@
+        uint256 totalPremiums,
+        uint256 totalOptions
+    );
+
+    /**
+     * @notice Emitted when a buyer is whitelisted for purchasing options
+     * @param buyer The whitelisted buyer
+     */
+    event BuyerWhitelisted(address indexed buyer);
+
+    /**
+     * @notice Emitted when a buyer is removed from the whitelist for purchasing options
+     * @param buyer The blacklisted buyer
+     */
+    event BuyerBlacklisted(address indexed buyer);
+
+    /**
+     * @notice Emitted when the ceiling price for a vault is updated
+     * @param vault The vault
+     * @param ceilingPrice The new ceiling price
+     */
+    event CeilingPriceUpdated(address indexed vault, uint256 ceilingPrice);
+
+    /************************************************
+     *  CONSTRUCTOR
+     ***********************************************/
+
+    /**
+     * @param _gnosisEasyAuction Gnosis Easy Auction contract
+     */
+    constructor(address _gnosisEasyAuction) {
+        gnosisEasyAuction = IGnosisAuction(_gnosisEasyAuction);
+    }
+
+    /************************************************
+     *  QUEUE OPERATIONS
+     ***********************************************/
+
+    /**
+     * @notice Create a request to purchase options from a vault at the auction settlement price
+     * @dev The buyer must be whitelisted to prevent the purchase queue from getting griefed (since sellToBuyers()
+     *  iterates through it). This function transfers the premiums for the options from the buyer at the ceiling
+     *  price (maximum price the buyer has to pay), however when the options are sold the buyer only pays the
+     *  auction settlement price and the leftover premiums are transferred back to the buyer. This function will
+     *  revert after the vault calls allocateOptions. New purchases can be made after the vault calls sellToBuyers().
+     *  The requests on the purchased queue are filled FIFO. Any unfilled/partially filled requests are refunded
+     *  their premiums, this can occur when the vault allocates less options than there are on the queue.
+     * @param vault The vault to purchase options from
+     * @param optionsAmount Amount of options requested
+     * @return premiums Amount of premiums transferred from the buyer
+     */
+    function requestPurchase(address vault, uint256 optionsAmount)
+        external
+        override
+        returns (uint256)
+    {
+        require(whitelistedBuyer[msg.sender], "Buyer not whitelisted");
+        uint256 _ceilingPrice = ceilingPrice[vault];
+        require(_ceilingPrice != 0, "Invalid vault");
+        // This prevents new purchase requested after the vault has set its allocation
+        require(vaultAllocatedOptions[msg.sender] == 0, "Vault allocated");
+        require(optionsAmount != 0, "!optionsAmount");
+
+        // premiums = optionsAmount * ceilingPrice
+        uint256 premiums =
+            optionsAmount.mul(_ceilingPrice).div(10**Vault.OTOKEN_DECIMALS);
```

Comment:
Also, shouldn't this be divided by the asset's decimal? Like:

```solidity
uint premiums = (optionsAmount * ceilingPrice) / assetDecimals
```

So for buying 1 put option with ceiling price of 15 USDC, it would be:

(1*10**8 * 15*10**6) / (10**6)

Reply:
Since the ceilingPrice is denominated in the asset's decimals already, it should be divided by the oToken's decimal

Example: Buying 10 put options with ceiling price of 15 USDC
`10e8 * 15e6 / 1e8 = 150e6 (150000000)`